### PR TITLE
fix: autosave chat controls on change

### PR DESCRIPTION
**Pull request by Algorithm5838**
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) to discuss your idea/fix with the community before creating a pull request, and describe your changes before submitting a pull request.

This is to ensure large feature PRs are discussed with the community first, before starting work on it. If the community does not want this feature or it is not relevant for Open WebUI as a project, it can be identified in the discussion before working on the feature and submitting the PR.

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Verify that the pull request targets the `dev` branch. **PRs targeting `main` will be immediately closed.**
- [x] **Description:** Provide a concise description of the changes made in this pull request down below.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [x] **Documentation:** Add docs in [Open WebUI Docs Repository](https://github.com/open-webui/docs). Document user-facing behavior, environment variables, public APIs/interfaces, or deployment steps.
- [x] **Dependencies:** Are there any new or upgraded dependencies? If so, explain why, update the changelog/docs, and include any compatibility notes. Actually run the code/function that uses updated library to ensure it doesn't crash.
- [x] **Testing:** Perform manual tests to **verify the implemented fix/feature works as intended AND does not break any other functionality**. Include reproducible steps to demonstrate the issue before the fix. Test edge cases (URL encoding, HTML entities, types). Take this as an opportunity to **make screenshots of the feature/fix and include them in the PR description**.
- [x] **Agentic AI Code:** Confirm this Pull Request is **not written by any AI Agent** or has at least **gone through additional human review AND manual testing**. If any AI Agent is the co-author of this PR, it may lead to immediate closure of the PR.
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Design & Architecture:** Prefer smart defaults over adding new settings; use local state for ephemeral UI logic. Open a Discussion for major architectural or UX changes.
- [x] **Git Hygiene:** Keep PRs atomic (one logical change). Clean up commits and rebase on `dev` to ensure no unrelated commits (e.g. from `main`) are included. Push updates to the existing PR branch instead of closing and reopening.
- [x] **Title Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

Editing chat controls (system prompt, params, files) updates frontend state immediately but never persists it. Changing a control and navigating away or refreshing before sending a message silently discards the edit.

This adds a debounced autosave in `Chat.svelte` that persists `params` and `chatFiles` to the chat record within 400ms of any genuine change from the controls panel.

### Added

- [List any new features, functionalities, or additions]

### Changed

- [List any changes, updates, refactorings, or optimizations]

### Deprecated

- [List any deprecated functionality or features that have been removed]

### Removed

- [List any removed features, files, or functionalities]

### Fixed

- Chat control changes (system prompt, params, file dismissals) are now autosaved and survive refresh or navigation without a message send.

### Security

- [List any new or updated security-related changes, including vulnerability fixes]

### Breaking Changes

- **BREAKING CHANGE**: [List any breaking changes affecting compatibility or functionality]

---

### Additional Information

- `Chat.svelte`: reactive watcher using `equal` (already imported from `fast-deep-equal`) to detect value changes and schedule a debounced `updateChatById` sending only `params` and `files`. Timer is cleared on navigation and component destroy to prevent stale writes.
- No new dependencies or API routes. Reuses `updateChatById`; the backend merges partials so history and title are unaffected.
- Save is blocked during initial load hydration (`loading`), before `$chatId` is set, and for temporary chats, consistent with all other save paths in the file.
- 86df8bf27 (now in dev) saves params/files when a new chat gets its ID. This covers the other case: control changes on existing chats, which skip that path.

### Screenshots or Videos

- [Attach any relevant screenshots or videos demonstrating the changes]

### Contributor License Agreement

- [x] By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](https://github.com/open-webui/open-webui/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

> [!NOTE]
> Deleting the CLA section will lead to immediate closure of your PR and it will not be merged in.

### CODE REVIEW

**Comment by Algorithm5838** [comment]
@pr-validator-bot review

**Comment by Algorithm5838** [comment]
Both findings addressed in the latest push.

Fast navigation: `navigateHandler` now flushes the pending save before canceling the timer, while `$chatId` still holds the outgoing chat.

Silent failures: The debounced save calls `toast.error` on rejection. The navigation flush only logs the error since toasting about a previous chat after navigating away would be confusing.

**Comment by nehaaprasad** [comment]
### AI Code Trust — ship readiness

**Score:** 95 · **Verdict:** RISKY · **Model:** deterministic+groq-v1

Notable issues found; review before shipping. Unvalidated user inputs are being passed to updateChatById function and null checks are missing for certain variables.

**Top issues:**
- **security** (high) — line 2739 passes user-controlled `params` and `chatFiles` into `updateChatById`, which can lead to unauthorized chat modifications
  _at_ [`src/lib/components/chat/Chat.svelte:2739`](https://github.com/open-webui/open-webui/blob/17edb99d8f8d1521940d9e12f755d4a3fac6e1af/src/lib/components/chat/Chat.svelte#L2739)
- **logic** (medium) — line 182 checks `!equal(params, oldParams)` but does not check if `params` is null or undefined, which can lead to a runtime error
  _at_ [`src/lib/components/chat/Chat.svelte:182`](https://github.com/open-webui/open-webui/blob/17edb99d8f8d1521940d9e12f755d4a3fac6e1af/src/lib/components/chat/Chat.svelte#L182)

**Comment by nehaaprasad** [comment]
### AI Code Trust — ship readiness

**Score:** 100 · **Verdict:** SAFE · **Model:** deterministic+groq-v1

No blocking issues detected by automated checks. Untrusted input from localStorage.token is passed to updateChatById without validation.

**Comment by nehaaprasad** [comment]
### AI Code Trust — ship readiness

**Score:** 93 · **Verdict:** RISKY · **Model:** deterministic+groq-v1

Notable issues found; review before shipping. Unvalidated user input is passed into `updateChatById`

**Top issues:**
- **security** (high) — Chat.svelte:2740 passes user-controlled `params` into `updateChatById` without validation
  _at_ [`src/lib/components/chat/Chat.svelte:2740`](https://github.com/open-webui/open-webui/blob/17edb99d8f8d1521940d9e12f755d4a3fac6e1af/src/lib/components/chat/Chat.svelte#L2740)
- **security** (high) — Chat.svelte:2741 passes user-controlled `chatFiles` into `updateChatById` without validation
  _at_ [`src/lib/components/chat/Chat.svelte:2741`](https://github.com/open-webui/open-webui/blob/17edb99d8f8d1521940d9e12f755d4a3fac6e1af/src/lib/components/chat/Chat.svelte#L2741)

File: src/lib/components/chat/Chat.svelte [hunk 1]
```diff
@@ -2708,6 +2733,15 @@
 		return _chatId;
 	};
 
+	const saveControlsHandler = async () => {
+		const _chatId = $chatId;
+		if (!_chatId || $temporaryChatEnabled || loading) return;
+		await updateChatById(localStorage.token, _chatId, {
+			params: params,
```
Comment (nehaaprasad) [inline src/lib/components/chat/Chat.svelte:2744]: **security** · high

Chat.svelte:2740 passes user-controlled `params` into `updateChatById` without validation

File: src/lib/components/chat/Chat.svelte [hunk 2]
```diff
@@ -2708,6 +2733,15 @@
...
+			files: chatFiles
```
Comment (nehaaprasad) [inline src/lib/components/chat/Chat.svelte:2745]: **security** · high

Chat.svelte:2741 passes user-controlled `chatFiles` into `updateChatById` without validation

**Comment by Algorithm5838** [comment]
Also fixed in the latest push: navigating to Settings or Admin unmounts the component without going through `navigateHandler`, silently dropping any pending save. The cleanup block now flushes it first.

**Comment by Algorithm5838** [comment]
It addresses https://github.com/open-webui/open-webui/discussions/23996

**Comment by Classic298** [comment]
is this still needed? it was addressed in dev i believe. please pull dev branch and test thanks

**Comment by Algorithm5838** [comment]
86df8bf27 covers new chats when they get their ID, but control changes on existing chats skip that path. This PR fills that gap.

**Comment by Classic298** [comment]
@owui-terminator review

**Comment by Algorithm5838** [comment]
1. clearTimeout null guard: Fixed. Wrapped in `if (controlsSaveTimer !== null)` to match the other two call sites.

2. Snapshot burned before save succeeds: By design. A failed save shows a toast; editing again will trigger another save. Explicit retry state adds complexity for an edge case.

3. Concurrent out-of-order saves: Not realistic. API calls complete well within 400ms, so a stale request can't realistically outlast a newer one.

4. Destroy-time flush on refresh/tab-close: `onDestroy` only fires on SPA navigation, which the `navigateHandler` flush already covers. Covering browser unload properly needs a `beforeunload` listener, which is out of scope.

5. Naming: Happy to rename to `lastObservedParams`/`lastObservedChatFiles` if preferred.

**Comment by Classic298** [comment]
yes please ensure naming is elegant for variables

**Comment by Algorithm5838** [comment]
Renamed to `lastObservedParams`/`lastObservedChatFiles`.

**Comment by tjbck** [comment]
Addressed in dev!
